pyomo/contrib/mindtpy/outer_approximation.py

```diff
@@ -0,0 +1,445 @@
+        if (config.add_no_good_cuts or config.use_tabu_list) and not self.should_terminate and config.add_regularization is None:
+            self.fix_dual_bound(config, last_iter_cuts)
+        config.logger.info(
+            ' ===============================================================================================')
+
+    def check_config(self):
+        config = self.config
+        if config.init_strategy is None:
+            config.init_strategy = 'rNLP'
+        if config.add_regularization is not None:
+            if config.add_regularization in {'grad_lag', 'hess_lag', 'hess_only_lag', 'sqp_lag'}:
+                config.calculate_dual_at_solution = True
+            if config.regularization_mip_threads == 0 and config.threads > 0:
+                config.regularization_mip_threads = config.threads
+                config.logger.info(
+                    'Set regularization_mip_threads equal to threads')
+            if config.single_tree:
+                config.add_cuts_at_incumbent = True
+                # if no method is activated by users, we will use use_bb_tree_incumbent by default
+                if not (config.reduce_level_coef or config.use_bb_tree_incumbent):
+                    config.use_bb_tree_incumbent = True
+            if config.mip_regularization_solver is None:
+                config.mip_regularization_solver = config.mip_solver
+        if config.single_tree:
+            config.logger.info('Single-tree implementation is activated.')
+            config.iteration_limit = 1
+            config.add_slack = False
+            if config.mip_solver not in {'cplex_persistent', 'gurobi_persistent'}:
+                raise ValueError("Only cplex_persistent and gurobi_persistent are supported for LP/NLP based Branch and Bound method."
+                                "Please refer to https://pyomo.readthedocs.io/en/stable/contributed_packages/mindtpy.html#lp-nlp-based-branch-and-bound.")
+            if config.threads > 1:
+                config.threads = 1
+                config.logger.info(
+                    'The threads parameter is corrected to 1 since lazy constraint callback conflicts with multi-threads mode.')
+        if config.heuristic_nonconvex:
+            config.equality_relaxation = True
+            config.add_slack = True
+        if config.equality_relaxation:
+            config.calculate_dual_at_solution = True
+        if config.init_strategy == 'FP' or config.add_regularization is not None:
+            config.move_objective = True
+        if config.add_regularization is not None:
+            if config.add_regularization in {'level_L1', 'level_L_infinity', 'grad_lag'}:
+                self.regularization_mip_type = 'MILP'
+            elif config.add_regularization in {'level_L2', 'hess_lag', 'hess_only_lag', 'sqp_lag'}:
+                self.regularization_mip_type = 'MIQP'
+        _MindtPyAlgorithm.check_config(self)
+
+
+    def initialize_mip_problem(self):
```

Comment:
This looks almost identical to the ECP implementation, and maybe others. Could you centralize the common parts in the base class or somewhere else?

Reply:
We can centralize the common parts in the base class. However, if we do so, we need to introduce several `if else` logic, which we are trying to avoid.

Reply:
We would be open to any suggestion, though, there is a fair share of repetition in the code, but the methods are pretty similar to each other. Any recommendation would be welcomed!

Reply:
Personally, I definitely prefer a little repeated code over introducing if-else's checking the algorithm in the base class. And I'm also not opposed to every algorithm having its own `solve` method, but it almost seems like you don't need it: Your solve methods are so similar right now that I think it makes sense to have a generic one for at least most of the algorithms, maybe with a few callbacks whose implementations you could define in the algorithm-specific classes where you do special things. For example, the only thing difference that I can see between ECP's solve and GOA's is lines 110-115 in `global_outer_approximation.py`. So maybe you include a `after_objective_processed` callback that is a no-op by default, but implements the adjustments you make in GOA.

I'm also fine with them all defining their own iteration loops--some repeated code seems worth making it easier to tell what each algorithm does. But I guess I would prefer those be very short methods so that it's easier to see the differences, and so that if you change something in the logic later, you don't have to go find it in 5 different files. So for example, maybe breaking out the logic for when not using a single-tree in GOA would make it easier to see what's the same and what's different.

So, I'm rambling but I guess my thoughts are:

1. You could consider a base class implementation of the similar stuff (`solve` and `iteration_loop`, and maybe `initialize_mip_problem`) that uses callbacks for the (slight) differences in algorithms. You have a better sense than me of whether this is a good design for what you intend to implement in the future.
2. You could ignore me for now (or forever)--I'm partly just playing devil's advocate, but I will say that when I reviewed it, I spent a lot of time with two algorithms open, trying to find the differences in implementation hidden in a sea of code I had already looked at.
3. Overall, I don't mean for this to hold up this PR--it's up to @bernalde and @ZedongPeng what design you like. But if it were up to me, I would prefer pretty much anything over if-else's checking for the algorithm all the time! :)